### PR TITLE
modified jni logic(List-> Array)

### PR DESCRIPTION
**Pull request by pluris**
In JNI code, it has been modified from List to Array.
If there are need to be fix,  please comment.

Issue : https://github.com/ros2-java/ros2_java/issues/149

### CODE REVIEW

**Comment by jacobperron** [comment]
Thanks for looking into this :slightly_smiling_face: 

Do you have more specifics on where the performance impact is coming from?

From a usability perspective, it might be better to continue to use `List` at the higher layer (e.g. `Node`, `AsyncParametersClient`, etc) if we the performance gains are strictly related to the changes to JNI code.

**Comment by pluris** [comment]
@jacobperron 
In task, for example, I want to continuously transmit image data, but there was no problem in the Ros2 cpp application, but in Ros2Java, there was a serious performance degradation.
Because of this problem, I am using it by changing to Array.
I agree that List is better for usability. However, I am concerned about List performance. If you have any other ideas, please suggest.

**Comment by jacobperron** [comment]
To clarify my previous comment, I do think that switching to Array in the C code makes sense, but my question is if it is necessary to also change all of the API in the Java client code? For example, I wouldn't think that changing the types used by the parameters API would have a significant performance impact on publishing and subscribing to image topics.

I plan to investigate the performance issues myself and will give a more informed review later.

File: rosidl_generator_java/resource/msg.cpp.em [hunk 1]
```diff
@@ -339,62 +348,74 @@ jobject @(underscore_separated_type_name)__convert_to_java(@(msg_normalized_type
...
-      _jlist_@(member.name)_element = env->NewStringUTF(_ros_@(member.name)_element.data);
-@[        else]@
-      _jlist_@(member.name)_element = env->NewString(_ros_@(member.name)_element.data, _ros_@(member.name)_element.size);
-@[        end if]@
-    }
+@[      if isinstance(member.type.value_type, AbstractString)]@
+      _jarray_@(member.name)_element = env->NewStringUTF(_ros_@(member.name)_element.data);
 @[      else]@
-    jobject _jlist_@(member.name)_element = env->NewObject(
-      _j@(normalized_type)_class_global, _j@(normalized_type)_constructor_global, _ros_@(member.name)_element);
```
Comment (jacobperron) [inline rosidl_generator_java/resource/msg.cpp.em:366]: IIUC, the performance issue is coming from creating new Java objects for each element in the list.

**Comment by jacobperron** [comment]
I haven't found a way to improve performance while continuing to use the `List` type.
Maybe it is acceptable to completely switch to array types given the performance gains. This means that users will no longer be able to change the size of ROS message members with "list" type, for example:

```java
// Before
List<Byte> image_data = message.getData();
image_data.get(0);  // access
image_data.add((byte) 200);  // modify message

// After
byte[] image_data = message.getData();
image_data[0];  // access
// No easy way to modify message in place, must construct a new array
int[] new_image_data = Arrays.copyOf(image_data, image_data.length + 1);
new_image_data[arr.length] = (byte) 200;
```

But, perhaps not being able to change the size of a ROS message in-place is a good thing. It might avoid bugs where the user modifies the returned list and unintentionally modifies the member of the message.

@esteve Any thoughts on this? I think we really do need to address this performance issue.

**Comment by jacobperron** [comment]
I think a good option is to switch to arrays by default and offer a convenience method for getting List types. I've adapting this PR and made some additional changes: https://github.com/osrf/ros2_java/tree/jacob/list_to_array

I'll adapt it for `dashing` and open a PR when it's ready.

**Comment by jacobperron** [comment]
Re-applying this change in https://github.com/ros2-java/ros2_java/pull/165.

**Pull request closed without merging**